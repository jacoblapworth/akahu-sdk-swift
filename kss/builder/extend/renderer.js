const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
