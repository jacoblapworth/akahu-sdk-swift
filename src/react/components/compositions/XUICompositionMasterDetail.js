import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import XUIGridAreaDetail from './XUIGridAreaDetail';
import XUIGridAreaMaster from './XUIGridAreaMaster';

import baseCompositionClass from './helpers';

export default class XUICompositionMasterDetail extends PureComponent {
	render() {
		const {
			master,
			detail,
			className,
			isInfinite,
		} = this.props;

		const compositionClasses = cn(
			baseCompositionClass,
			`${baseCompositionClass}-masterdetail`,
			!isInfinite && `${baseCompositionClass}-is-finite`,
			className,
		);

		return (
			<div className={compositionClasses}>
				<XUIGridAreaMaster>
					{master}
				</XUIGridAreaMaster>
				<XUIGridAreaDetail>
					{detail}
				</XUIGridAreaDetail>
			</div>
		);
	}
}

XUICompositionMasterDetail.propTypes = {
	className: PropTypes.string,

	/**
	 * Nav content or controls for detail content
	 */
	master: PropTypes.element.isRequired,
	/**
	 * Main content
	 */
	detail: PropTypes.element.isRequired,
	/**
	 * Determines whether the main content takes full width of page
	 */
	isInfinite: PropTypes.bool,

};
