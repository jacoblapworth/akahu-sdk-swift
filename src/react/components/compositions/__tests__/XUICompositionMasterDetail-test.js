import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { axe, toHaveNoViolations } from 'jest-axe';
import renderer from 'react-test-renderer';

import XUICompositionMasterDetail from '../XUICompositionMasterDetail';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

describe('<XUICompositionMasterDetail>', () => {
  it('renders basic example', () => {
    const wrapper = renderer.create(
      <XUICompositionMasterDetail master={<div />} detail={<div />} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should do nothing with children', () => {
    const wrapper = renderer.create(
      <XUICompositionMasterDetail master={<div />} detail={<div />}>
        Hello
      </XUICompositionMasterDetail>,
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should include custom class and toggle off default classes, if specified', () => {
    const wrapper = renderer.create(
      <XUICompositionMasterDetail
        master={<div />}
        detail={<div />}
        className="master-detail"
        isInfinite={true}
        hasAutoSpaceAround={false}
        hasGridGap={false}
        hasAutoColumnWidths={false}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should include layout restriction class, if specified', () => {
    const wrapper = renderer.create(
      <XUICompositionMasterDetail master={<div />} detail={<div />} retainWidth="small" />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass accessibility testing', async () => {
    const wrapper = mount(<XUICompositionMasterDetail detail={<div />} master={<div />} />);
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });
});
