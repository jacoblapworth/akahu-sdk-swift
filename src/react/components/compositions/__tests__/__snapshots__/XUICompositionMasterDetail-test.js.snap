// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUICompositionMasterDetail> renders basic example 1`] = `
<div
  className="xui-composition xui-composition-masterdetail xui-composition-is-finite"
>
  <div
    className="xui-gridarea-master"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;

exports[`<XUICompositionMasterDetail> should do nothing with children 1`] = `
<div
  className="xui-composition xui-composition-masterdetail xui-composition-is-finite"
>
  <div
    className="xui-gridarea-master"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;

exports[`<XUICompositionMasterDetail> should include layout restriction class, if specified 1`] = `
<div
  className="xui-composition xui-composition-masterdetail xui-composition-is-finite xui-composition-masterdetail-retain-small"
>
  <div
    className="xui-gridarea-master"
  >
    <div />
  </div>
  <div
    className="xui-gridarea-detail"
  >
    <div />
  </div>
</div>
`;
